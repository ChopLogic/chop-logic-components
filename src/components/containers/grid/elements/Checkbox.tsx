import React from 'react';
import { useElementIds } from 'hooks/use-element-ids';

import ChopLogicLabel from 'components/inputs/_common/label/Label';
import CheckboxCheckedIcon from 'components/misc/icon/elements/CheckboxChecked';
import CheckboxUncheckedIcon from 'components/misc/icon/elements/CheckboxUnchecked';

import { StyledGridCheckboxInput, StyledGridCheckboxWrapper } from '../Grid.styled';

type ChopLogicGridCheckboxProps = React.InputHTMLAttributes<HTMLInputElement> & {
  label: string;
};

const GridCheckbox: React.FC<ChopLogicGridCheckboxProps> = ({
  name,
  label,
  disabled,
  required = false,
  onChange,
  checked = false,
  ...props
}) => {
  const { elementId } = useElementIds(props?.id);

  return (
    <StyledGridCheckboxWrapper $disabled={!!disabled}>
      <StyledGridCheckboxInput
        id={elementId}
        name={name}
        type='checkbox'
        disabled={disabled}
        required={required}
        checked={checked}
        onChange={onChange}
      />
      <ChopLogicLabel
        label={label}
        required={required}
        inputId={elementId}
        isTextHidden={true}
        iconPosition='left'
        icon={checked ? <CheckboxCheckedIcon /> : <CheckboxUncheckedIcon />}
      />
    </StyledGridCheckboxWrapper>
  );
};

export default GridCheckbox;
