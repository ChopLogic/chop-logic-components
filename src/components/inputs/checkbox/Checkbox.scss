@import 'styles/global.scss';

.cl-checkbox {
  display: flex;
  gap: var(--small-gap);
  padding: var(--small-gap) 0;
  position: relative;
  overflow: hidden;

  &__input {
    position: absolute;
    z-index: -1;
    opacity: 0;
    top: 4px;
    left: 0;
    margin: 0;

    &:focus-visible + label {
      outline: var(--focus-outline);
      outline-offset: 2px;
    }
  }

  &__label {
    cursor: pointer;
    align-items: stretch !important;
    font-size: 1rem !important;

    &::before {
      color: var(--cl-color-celtic-blue);
    }

    &:hover,
    &:active {
      text-shadow: var(--text-shadow);
    }
  }

  &_disabled {
    pointer-events: none;
    filter: grayscale(80%);
    opacity: 0.6;
  }
}
