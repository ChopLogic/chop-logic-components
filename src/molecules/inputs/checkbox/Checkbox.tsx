import React from 'react';
import { useElementIds } from '@hooks';
import CheckboxCheckedIcon from '@assets/icons/svg/CheckboxCheckedIcon.tsx';
import CheckboxUncheckedIcon from '@assets/icons/svg/CheckboxUncheckedIcon.tsx';
import { ChopLogicCheckboxProps } from '@models';
import { useChopLogicCheckboxController } from './Checkbox.controller.ts';
import styles from './Checkbox.module.scss';
import { getClassName } from '@utils';
import { Label } from '@atoms';

const ChopLogicCheckbox: React.FC<ChopLogicCheckboxProps> = ({
  name,
  label,
  disabled,
  noLabel,
  required = false,
  iconPosition = 'left',
  defaultChecked,
  onChange,
  id,
  className,
  ...rest
}) => {
  const { handleChange, checked } = useChopLogicCheckboxController({ name, defaultChecked, onChange });
  const { elementId } = useElementIds(id);
  const checkboxClass = getClassName([
    styles.checkbox,
    className,
    {
      [styles.checkbox__disabled]: !!disabled,
      [styles.checkbox__checked]: checked,
    },
  ]);

  return (
    <div {...rest} className={checkboxClass}>
      <input
        id={elementId}
        name={name}
        type='checkbox'
        disabled={disabled}
        required={required}
        checked={checked}
        onChange={handleChange}
        aria-label={noLabel ? label : undefined}
      />
      <Label
        label={label}
        required={required}
        inputId={elementId}
        isTextHidden={noLabel}
        icon={checked ? <CheckboxCheckedIcon /> : <CheckboxUncheckedIcon />}
        iconPosition={iconPosition}
      />
    </div>
  );
};

export default ChopLogicCheckbox;
