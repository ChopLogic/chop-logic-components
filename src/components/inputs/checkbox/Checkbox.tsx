import React from 'react';
import { useElementIds } from 'hooks/use-element-id';

import CheckboxCheckedIcon from 'components/misc/icon/elements/CheckboxChecked';
import CheckboxUncheckedIcon from 'components/misc/icon/elements/CheckboxUnchecked';
import ChopLogicLabel from 'components/misc/label/Label';

import { StyledCheckboxInput, StyledCheckboxWrapper } from './Checkbox.styled';
import { useChopLogicCheckboxController } from './helpers';

export type ChopLogicCheckboxProps = React.InputHTMLAttributes<HTMLInputElement> & {
  name: string;
  label: string;
  isLabelHidden?: boolean;
  iconPosition?: 'left' | 'right';
};

const ChopLogicCheckbox: React.FC<ChopLogicCheckboxProps> = ({
  name,
  label,
  disabled,
  isLabelHidden,
  required = false,
  iconPosition = 'left',
  defaultChecked,
  onChange,
  ...props
}) => {
  const { handleChange, checked } = useChopLogicCheckboxController({ name, defaultChecked, onChange });
  const { elementId } = useElementIds(props?.id);

  return (
    <StyledCheckboxWrapper
      $disabled={!!disabled}
      className={props?.className}
      tabIndex={props?.tabIndex}
      style={props?.style}
      title={props?.title}
    >
      <StyledCheckboxInput
        id={elementId}
        name={name}
        type='checkbox'
        disabled={disabled}
        required={required}
        checked={checked}
        onChange={handleChange}
        aria-label={isLabelHidden ? label : undefined}
      />
      <ChopLogicLabel
        label={label}
        required={required}
        inputId={elementId}
        isTextHidden={isLabelHidden}
        icon={checked ? <CheckboxCheckedIcon /> : <CheckboxUncheckedIcon />}
        iconPosition={iconPosition}
      />
    </StyledCheckboxWrapper>
  );
};

export default ChopLogicCheckbox;
