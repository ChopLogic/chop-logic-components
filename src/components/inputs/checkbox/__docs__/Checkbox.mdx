import { Canvas, Meta } from '@storybook/blocks';
import * as CheckboxStories from './Checkbox.stories';

<Meta of={CheckboxStories} title='Checkbox' />

# ChopLogicCheckbox

Styled checkbox component with an optional label and floating checkbox icon. Can be used as input for ChopLogicForm.

### Default View

<Canvas of={CheckboxStories.LeftIconCheckbox} />

### Usage

```tsx
import { ChopLogicCheckbox } from 'chop-logic-components';

const Example = () => {
  return (
    <ChopLogicCheckbox
      id='terms-and-conditions-checkbox'
      name='isTermsAccepted'
      label='Accept Terms and Conditions'
      iconPosition='left'
      required
      onChange={(e) => console.log('Checkbox state:', e.target.checked)}
      disabled={false}
    />
  );
};

export default Example;
```

### Props

| Prop               | Type                  | Description                                                                                                                                                | Default     | Required |
|--------------------|-----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------|----------|
| **name**           | `string`              | A string representing the name attribute of the component, used for form submissions.                                                                      | `undefined` | Yes      |
| **label**          | `string`              | A string that represents the label for the component.                                                                                                      | `undefined` | Yes      |
| **defaultChecked** | `boolean`             | A boolean indicating whether the checkbox is checked by default or not.                                                                                    | `false`     | No       |
| **onChange**       | `ChangeEventHandler`  | A function that is called when the checkbox state changes. It receives a change event.                                                                     | `undefined` | No       |
| **noLabel**        | `boolean`             | A boolean indicating whether the label should be hidden or not. The label prop is still required in this case, because it's used for aria-label attribute. | `false`     | No       |
| **iconPosition**   | `'left'` or `'right'` | The place where the checkbox icon will be rendered (left or right).                                                                                        | `'left'`    | No       |
| **disabled**       | `boolean`             | A boolean indicating whether the checkbox should be disabled. When disabled, the checkbox cannot be interacted with.                                       | `false`     | No       |
| **required**       | `boolean`             | A boolean indicating whether the checkbox is required in a form. When required, the checkbox must be checked before form submission.                       | `false`     | No       |
| **id**             | `string`              | The id global attribute defines an identifier (ID) which must be unique in the whole document.                                                             | `undefined` | No       |
| **className**      | `string`              | The class global attribute is a list of the classes of the element.                                                                                        | `undefined` | No       |
| **style**          | `CSSProperties`       | The style global attribute contains CSS styling declarations to be applied to the element.                                                                 | `undefined` | No       |
| **tabIndex**       | `number`              | The tabindex global attribute allows developers to make HTML elements focusable.                                                                           | `undefined` | No       |
| **title**          | `string`              | The title global attribute contains text representing advisory information related to the element it belongs to.                                           | `undefined` | No       |
| **theme**          | `'light'` or `'dark'` | The component's color theme.                                                                                                                               | `undefined` | No       |

### Types

```ts
export type ChopLogicCheckboxProps = CommonInputProps & {
  noLabel?: boolean;
  iconPosition?: 'left' | 'right';
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  defaultChecked?: boolean;
};

export type CommonInputProps = CommonComponentProps & {
  label: string;
  name: string;
  disabled?: boolean;
  required?: boolean;
};

export type CommonComponentProps = {
  id?: string;
  className?: string;
  style?: React.CSSProperties;
  tabIndex?: number;
  title?: string;
  theme?: 'dark' | 'light';
};
```
