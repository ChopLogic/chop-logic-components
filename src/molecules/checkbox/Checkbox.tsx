import { Label } from '@atoms';
import { IconName } from '@enums';
import { useElementIds } from '@hooks';
import { CheckboxProps } from '@models';
import { getClassName } from '@utils';
import { FC } from 'react';

import { useCheckboxController } from './Checkbox.controller';
import styles from './Checkbox.module.scss';

const Checkbox: FC<CheckboxProps> = ({
  name,
  label,
  disabled,
  noLabel,
  required = false,
  iconPosition = 'left',
  defaultChecked,
  onChange,
  id,
  className,
  ...rest
}) => {
  const { handleChange, checked } = useCheckboxController({ name, defaultChecked, onChange });
  const { elementId } = useElementIds(id);
  const checkboxClass = getClassName([
    styles.checkbox,
    className,
    {
      [styles.checkbox__disabled]: !!disabled,
      [styles.checkbox__checked]: checked,
    },
  ]);

  return (
    <div {...rest} className={checkboxClass}>
      <input
        id={elementId}
        name={name}
        type='checkbox'
        disabled={disabled}
        required={required}
        checked={checked}
        onChange={handleChange}
        aria-label={noLabel ? label : undefined}
      />
      <Label
        label={label}
        required={required}
        inputId={elementId}
        isTextHidden={noLabel}
        icon={checked ? IconName.CheckboxChecked : IconName.CheckboxUnchecked}
        iconPosition={iconPosition}
      />
    </div>
  );
};

export default Checkbox;
