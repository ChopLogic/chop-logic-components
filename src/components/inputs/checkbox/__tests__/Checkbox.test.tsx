import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';

import Checkbox from '../Checkbox';
import userEvent from '@testing-library/user-event';

describe('ChopLogicTextInput tests:', () => {
  const testProps = {
    id: 'test-checkbox',
    name: 'testCheckbox',
    label: 'Test checkbox:',
  };

  it('should render the input correctly', () => {
    const { asFragment } = render(<Checkbox {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display the label', () => {
    render(<Checkbox {...testProps} />);
    expect(screen.getByLabelText(testProps.label)).toBeInTheDocument();
  });

  it('should be enabled by default', () => {
    render(<Checkbox {...testProps} />);
    expect(screen.getByRole('checkbox')).toBeEnabled();
  });

  it('could be disable by prop', () => {
    render(<Checkbox {...testProps} disabled />);
    expect(screen.getByRole('checkbox')).toBeDisabled();
  });

  it('should check and uncheck the checkbox on click', async () => {
    render(<Checkbox {...testProps} />);
    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
    await userEvent.click(checkbox);
    expect(checkbox).toBeChecked();
    await userEvent.click(checkbox);
    expect(checkbox).not.toBeChecked();
  });

  it('should call onChange handler on click', async () => {
    const mockOnChange = vi.fn();
    render(<Checkbox {...testProps} onChange={mockOnChange} />);
    const checkbox = screen.getByRole('checkbox');
    await userEvent.click(checkbox);
    expect(mockOnChange).toHaveBeenCalledOnce();
    expect(checkbox).toBeChecked();
  });

  it('should call onChange handler on Space key press', async () => {
    const mockOnChange = vi.fn();
    render(<Checkbox {...testProps} onChange={mockOnChange} />);
    const checkbox = screen.getByRole('checkbox');
    checkbox.focus();
    await userEvent.keyboard('[Space]');
    expect(mockOnChange).toHaveBeenCalledOnce();
    expect(checkbox).toBeChecked();
  });
});
