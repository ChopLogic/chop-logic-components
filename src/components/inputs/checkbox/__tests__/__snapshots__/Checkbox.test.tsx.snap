// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicCheckbox > should render the input correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-ghWlax iTqVyz"
  >
    <input
      id="test-checkbox-id"
      name="checkbox"
      type="checkbox"
    />
    <label
      class="sc-ivxoEo kewDTQ"
      for="test-checkbox-id"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M14 0H2C.9 0 0 .9 0 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V2c0-1.1-.9-2-2-2zm0 14H2V2h12v12z"
          fill="currentColor"
        />
      </svg>
      <span>
        Test label
      </span>
    </label>
  </div>
</DocumentFragment>
`;
